x/valset-pref/types/msgs_test.go

```diff
@@ -0,0 +1,137 @@
+
+	tests := []struct {
+		name       string
+		msg        types.MsgSetValidatorSetPreference
+		expectPass bool
+	}{
+		{
+			name: "proper msg",
+			msg: types.MsgSetValidatorSetPreference{
+				Delegator: addr1,
+				Preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDecWithPrec(5, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1jcr68jghzm24zwe78zuhz7xahua8429erxk7vm",
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1gqsr38e4zteekwr6kq5se5jpadafqmcfyz8jds",
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "duplicate validator msg",
+			msg: types.MsgSetValidatorSetPreference{
+				Delegator: addr1,
+				Preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDecWithPrec(6, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDecWithPrec(4, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1jcr68jghzm24zwe78zuhz7xahua8429erxk7vm",
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name: "invalid delegator",
+			msg: types.MsgSetValidatorSetPreference{
+				Delegator: invalidAddr,
+				Preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDec(1),
+					},
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name: "invalid validator address",
+			msg: types.MsgSetValidatorSetPreference{
+				Delegator: addr1,
+				Preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxay", // invalid address
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1jcr68jghzm24zwe78zuhz7xahua8429erxk7vm",
+						Weight:         sdk.NewDecWithPrec(2, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDecWithPrec(6, 1),
+					},
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name: "weights != 1",
+			msg: types.MsgSetValidatorSetPreference{
+				Delegator: addr1,
+				Preferences: []types.ValidatorPreference{
+					{
+						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
+						Weight:         sdk.NewDecWithPrec(5, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1jcr68jghzm24zwe78zuhz7xahua8429erxk7vm",
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+					{
+						ValOperAddress: "osmovaloper1gqsr38e4zteekwr6kq5se5jpadafqmcfyz8jds",
+						Weight:         sdk.NewDecWithPrec(3, 1),
+					},
+				},
+			},
+			expectPass: false,
+		},
```

Comment:
super nit: maybe we should have a case where weights are < 1 as well

Reply:
yup will add it